Created generic quick pick util class to support various use cases

gnana997 opened the pull request:
## Summary of Changes

<!-- Include a high-level overview of your implementation, including any alternatives you considered and items you'll address in follow-up PRs -->

- Added a generic createQuickPick utility class to handle multiple scenarios exposing a callback functions for required usecase.

## Any additional details or context that should be provided?

<!-- Behavior before/after, more technical details/screenshots, follow-on work that should be expected, links to discussions or issues, etc -->

- #1127 

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

##### Tests

- [x] Added new
- [ ] Updated existing
- [ ] Deleted existing

##### Other

- [ ] All new disposables (event listeners, views, channels, etc.) collected as  for eventual cleanup?
<!-- prettier-ignore -->
- [ ] Does anything in this PR need to be mentioned in the user-facing [CHANGELOG](https://github.com/confluentinc/vscode/blob/main/CHANGELOG.md) or [README](https://github.com/confluentinc/vscode/blob/main/public/README.md)?
- [ ] Have you validated this change locally by [packaging](https://github.com/confluentinc/vscode/blob/main/README.md#packaging-steps) and installing the extension `.vsix` file?
  ```shell
  gulp clicktest
  ```

gnana997 commented:
Hi @shouples , I went through the` localResource.ts` and I have updated the utils quick pick to customize the `createQuickPick` in `localResource.ts`. Can you check it once and let me know if I am missing something?
Thanks!

shouples commented:
Hey @gnana997, this is working nicely so far! Just one test failure running locally on MacOS 15.3 (arm64):
```
  862 passing (4s)
  2 pending
  1 failing
  1) QuickPick utils
       should create a QuickPick with correct options:
     AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:

[
  {
    label: 'Item 2'
  }
]

should loosely deep-equal

{
  label: 'Item 2'
}
```

My only suggestion for now would be to include `QuickPickItemWithValue` as the primary item type (instead of `QuickPickItem`) in `quickPickUtils`:
https://github.com/confluentinc/vscode/blob/main/src/quickpicks/types.ts, and to provide some examples in the PR description or JSDoc comments for how to use the function(s) for simple quickpick use-cases as well as more complicated/customized use-cases.

Thanks for helping out with this -- I'll try to provide a more thorough review next week.

gnana997 commented:
Hi @shouples , I have updated the failing test case and also updated the class to use QuickPickItemWithValue along with JSDoc comments on usage.

Should I start migrating functions with `showQuickPick` to use `showEnhancedQuickPick` created?

Thanks!

shouples commented on src/utils/quickPickUtils.test.ts:
```diff
@@ -0,0 +1,152 @@
+    sandbox = sinon.createSandbox();
+
+    // Create a mock for the QuickPick object
+    quickPickMock = {
```
@gnana997 were you able to run `gulp test` with just the `show`/`hide` methods stubbed? I don't know if we need to stub all the properties of the quickpick.

gnana997 replied:
Hi @shouples , Yeah they all run successfully. 
Is there any issue with the test cases?
Should I add another test case with only show/hide methods stubbed?

Thanks!

shouples commented:
/sem-approve

shouples commented on src/utils/quickPickUtils.ts:
```diff
@@ -0,0 +1,306 @@
+ * @param options Enhanced QuickPick options
+ * @returns A promise that resolves to the selected item(s) or undefined if canceled
+ */
+export async function showEnhancedQuickPick<T extends QuickPickItemWithValue<vscode.QuickPickItem>>(
```
Same type comment here for `QuickPickItemWithValue<vscode.QuickPickItem>`

shouples reviewed (changes requested):
This is looking really nice so far @gnana997! Let's handle migrating the rest of the quickpicks in other smaller PRs. I'll write up some follow-on issues once this PR is merged.

For now I just have a couple questions and suggestions for some cleanup/consolidation.

gnana997 commented on src/utils/quickPickUtils.ts:
```diff
@@ -0,0 +1,306 @@
+  return options?.canSelectMany ? selectedItems : selectedItems[0];
+}
+
+/**
+ * Creates and returns a QuickPick instance with enhanced functionality.
+ * This allows more direct control over the QuickPick compared to showEnhancedQuickPick.
+ *
+ * @param items The items to show in the QuickPick
+ * @param options Enhanced QuickPick options
+ * @returns The configured QuickPick instance
+ */
+export function createEnhancedQuickPick<T extends QuickPickItemWithValue<vscode.QuickPickItem>>(
```
Yeah I think so too. I will see if I missed any other used features in `createQuickPick` and I will try to add them here.

However I think as part of smaller issues migrating other `showQuickPicks` to use generic enhanced quick pick if I find something that is missing I can handle during that too.

Let me know.

Thanks!

gnana997 replied:
I have removed the `showEnhancedQuickPick` function and updated the test cases on `createEnhancedQuickPick`
Thanks!

gnana997 commented on src/utils/quickPickUtils.ts:
Got it. I will move them

gnana997 commented on src/quickpicks/localResources.ts:
```diff
@@ -79,62 +74,62 @@ export async function localResourcesQuickPick(
+  quickpick.show();
 
   // user clicked "Ok" or hit Enter to accept the selected items, if any
   const selectedItems: QuickPickItem[] = [];
   quickpick.onDidAccept(() => {
     logger.debug("selected items", { items: JSON.stringify(quickpick.selectedItems) });
     selectedItems.push(...quickpick.selectedItems);
     quickpick.hide();
```
Hi @shouples , I have added the onDidAccept as optional in the interface.

gnana997 commented on src/utils/quickPickUtils.ts:
````diff
@@ -0,0 +1,306 @@
+ * };
+ * ```
+ */
+export interface EnhancedQuickPickOptions<T extends QuickPickItemWithValue<vscode.QuickPickItem>>
````
Hi @shouples , I have updated it accordingly.
Thanks!

shouples commented on src/quickpicks/localResources.ts:
```diff
@@ -79,62 +74,62 @@ export async function localResourcesQuickPick(
-  );
+  });
+
+  // Show the QuickPick
+  quickpick.show();
```
What do you think of making this the default behavior for the new function? Ideally we shouldn't need to have the callers explicitly `.show()` or block until `.onDidHide` fires.

gnana997 replied:
Yeah, That makes sense too. I was thinking this will provide us more control on using the `genericQuickPick`. I have refactored the `genericQuickPick` now.
Thanks!

shouples commented on src/quickpicks/localResources.ts:
```diff
@@ -79,62 +74,62 @@ export async function localResourcesQuickPick(
+  // Show the QuickPick
+  quickpick.show();
 
   // user clicked "Ok" or hit Enter to accept the selected items, if any
   const selectedItems: QuickPickItem[] = [];
   quickpick.onDidAccept(() => {
     logger.debug("selected items", { items: JSON.stringify(quickpick.selectedItems) });
     selectedItems.push(...quickpick.selectedItems);
     quickpick.hide();
```
We can also safely remove this section now that the new function accepts a callback for `onDidAccept`

gnana997 replied:
Yeah, I have removed it.
Thanks!

shouples commented on src/quickpicks/localResources.ts:
```diff
@@ -79,62 +74,62 @@ export async function localResourcesQuickPick(
-    return;
   });
 
   // block until the quickpick is hidden
```
And we should probably remove this block as well if the new function blocks by default

shouples commented on src/quickpicks/utils/quickPickUtils.test.ts:
```diff
@@ -0,0 +1,222 @@
+      show: sandbox.stub(),
+      hide: sandbox.stub(),
+      dispose: sandbox.stub(),
+      onDidChangeSelection: createEventStub<readonly any[]>(),
```
I think explicitly setting up individual stubs for these (like `onDidChangeSelectionStub`) will be more straightforward. Then they be passed directly into the assertions without having to use patterns like `const stub = quickPickMock.onDidChangeSelection as sinon.SinonStub;`. What do you think?

gnana997 replied:
Yeah I created all the stubs individually and used them for mocking the quickpick.
Thanks!

shouples commented on src/quickpicks/localResources.ts:
```diff
@@ -79,72 +74,54 @@ export async function localResourcesQuickPick(
     },
-  );
-
-  // user clicked "Ok" or hit Enter to accept the selected items, if any
-  const selectedItems: QuickPickItem[] = [];
-  quickpick.onDidAccept(() => {
-    logger.debug("selected items", { items: JSON.stringify(quickpick.selectedItems) });
-    selectedItems.push(...quickpick.selectedItems);
-    quickpick.hide();
-    return;
-  });
```
Looks like this logic needs to be added to the `onDidAccept` option for `createEnhancedQuickPick` since hitting Enter or clicking OK doesn't do anything at the moment

gnana997 replied:
oh yeah, Sorry I missed that. I just added a default behavior for `onDidAccept` along with the support to accept the required behavior and updated the class. I have refactored test cases to check the selected Items too.
Thanks!

shouples commented on src/quickpicks/utils/quickPickUtils.ts:
```diff
@@ -0,0 +1,251 @@
+  quickPick.show();
+  return new Promise((resolve) => {
+    quickPick.onDidHide(() => {
+      // If we have pre-selected items and haven't captured any selections yet,
+      // use the pre-selected items
+      if (selectedItems.length === 0 && options?.selectedItems) {
+        selectedItems = [...options.selectedItems];
+      }
```
I noticed some strange behavior when clicking the ⚙️ icon (either against the Kafka item or the SR item) where it seems to continue the user flow as if clicking "OK" or hitting "Enter":

https://github.com/user-attachments/assets/643eee3f-8cc4-4930-9739-97e8c392871e

I think removing the section here and only assigning `selectedItems` during `onDidAccept` will fix this and keep the behavior consistent with what's currently on `main`.

shouples replied:
Thanks for taking care of those other fixes @gnana997! Looks like this is the last part needed before we can merge.

shouples reviewed (changes requested):
This is really close @gnana997! Just one thing left to fix and then I think this one is good to merge. 🙏

shouples commented on src/quickpicks/utils/quickPickUtils.ts:
```diff
@@ -0,0 +1,251 @@
+    quickPick.selectedItems = options.selectedItems;
+  }
+
+  // Set up event handlers
+  if (options?.onSelectionChange) {
+    quickPick.onDidChangeSelection((items: readonly T[]) => {
+      options.onSelectionChange?.(items, quickPick);
+    });
+  }
+
+  if (options?.onActiveItemChange) {
+    quickPick.onDidChangeActive((items: readonly T[]) => {
+      options.onActiveItemChange?.(items[0], quickPick);
+    });
+  }
```
One other thing I noticed: let's make sure we're keeping track of these VS Code `Disposable`s (and the others below this block):
```suggestion
  const disposables: Disposable[] = [];
  
  // Set up event handlers
  if (options?.onSelectionChange) {
    disposables.push(
      quickPick.onDidChangeSelection((items: readonly T[]) => {
        options.onSelectionChange?.(items, quickPick);
      })
    );
  }

  if (options?.onActiveItemChange) {
    disposables.push(
      quickPick.onDidChangeActive((items: readonly T[]) => {
        options.onActiveItemChange?.(items[0], quickPick);
      });
    )
  }
```
(This is new since the code on `main` wasn't properly cleaning them up, but we'll want to include this so other quickpicks using this won't have to worry about listener cleanup.)

shouples commented on src/quickpicks/utils/quickPickUtils.ts:
```diff
@@ -0,0 +1,251 @@
+      if (selectedItems.length === 0 && options?.selectedItems) {
+        selectedItems = [...options.selectedItems];
+      }
+      resolve({ quickPick, selectedItems });
```
And then we'll need something like this to dispose those event listeners from above:
```suggestion
      disposables.forEach(d => d.dispose());
      resolve({ quickPick, selectedItems });
```
